app/pb/events.proto

```diff
@@ -0,0 +1,304 @@
+syntax = "proto3";
+
+package org.couchers.api.events;
+
+import "google/protobuf/empty.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/wrappers.proto";
+
+service Events {
+  // This is a secure service: a user needs to be authenticated and not jailed to call functions here, refer to auth.proto and jail.proto
+
+  rpc CreateEvent(CreateEventReq) returns (Event) {
+    // Creates an event, only one instance, add more with ScheduleEvent
+  }
+
+  rpc ScheduleEvent(ScheduleEventReq) returns (Event) {
+    // Schedules a new occurence of an existing event
+  }
+
+  rpc UpdateEvent(UpdateEventReq) returns (Event) {
+    // Edit an event
+  }
+
+  rpc GetEvent(GetEventReq) returns (Event) {
+    // Gets an event
+  }
+
+  rpc ListEventOccurences(ListEventOccurencesReq) returns (ListEventOccurencesRes) {
+    // List all occurences of a recurring event
+  }
+
+  rpc ListEventAttendees(ListEventAttendeesReq) returns (ListEventAttendeesRes) {
+    // Paginated list of event attendees
+  }
+
+  rpc ListEventSubscribers(ListEventSubscribersReq) returns (ListEventSubscribersRes) {
+    // Paginated list of event subscribers
+  }
+
+  rpc ListEventOrganizers(ListEventOrganizersReq) returns (ListEventOrganizersRes) {
+    // Paginated list of event organizers
+  }
+
+  rpc TransferEvent(TransferEventReq) returns (Event) {
+    // Transfer an event you own to either a group or a community
+  }
+
+  rpc SetEventSubscription(SetEventSubscriptionReq) returns (Event) {
+    // Subscribe to get notifications about all occurences of an event
+  }
+
+  rpc SetEventAttendance(SetEventAttendanceReq) returns (Event) {
+    // RSVP/change going status for an occurence of an event
+  }
+
+  rpc InviteEventOrganizer(InviteEventOrganizerReq) returns (google.protobuf.Empty) {
+    // Invite another user to become an organizer of this event
+  }
+
+  rpc RemoveEventOrganizer(RemoveEventOrganizerReq) returns (google.protobuf.Empty) {
+    // Remove yourself as an event organizer
+  }
+
+  rpc ListMyEvents(ListMyEventsReq) returns (ListMyEventsRes) {
+    // Lists all subscribed, attending
+  }
+
+  // TODO: DownloadCalendarInvite
```

Comment:
Client-side

Reply:
My idea was to generate the file (an icalendar, `.ics`) on the backend and just send that to the frontends. It's a standard format for describing events, and I think much better to do that in the backend instead of having to replicate it to each frontend implementation. There will probably also be some details that are just easier to do in the backend (e.g. tz support/conversions).

Reply:
Or we can provide a javascript function to generate an ics file from GetEvent response, that all frontends can use.

Reply:
Not all frontends will be written in javascript.

I am also hoping to have better integration with calendar software eventually. It's possible to add a URL that some software will update the event from, so we could potentially push new scheduled occurences directly to the user's calendar. This is not something we can do client side.

Also, client side icalendar handling libraries seem to be pretty crappy compared to https://pypi.org/project/icalendar/

Reply:
I disagree but both ways have their merits, so resolving conversation :)